### PR TITLE
core: implement path properties endpoint for time table v2

### LINKED ISSUE #6786: TS2: Add core path properties endpoint
## Description

Add an endpoint `POST` 

We should be able to query these properties:
- slopes
- gradients
- electrification (we should check if we depend on the rolling stock)
  - Contains neutral_sections
- geometry
- operational_points

## AC

- [ ] We should only return the queried properties

### PR DESCRIPTION
**Pull request by Erashin**
Path properties v2 core endpoint + editoast call.

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropEndpoint.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+            // Load infra
+            val infra = infraManager.getInfra(request.infra, request.expectedVersion, recorder)
+
+            val pathProps = makePathProps(infra.rawInfra, request.trackRanges)
+            val res = makePathPropResult(pathProps, request.rsSupportedElectrifications)
+
+            RsJson(RsWithBody(PathPropResult.adapterResult.toJson(res)))
+        } catch (ex: Throwable) {
+            // TODO: include warnings in the response
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropEndpoint.kt]: That todo is outdated, we don't really have warnings anymore

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    @Json(name = "expected_version") var expectedVersion: String,
+) {
+
+    companion object {
+        val adapter: JsonAdapter<PathPropRequest> =
+            Moshi.Builder()
+                .add(KotlinJsonAdapterFactory())
+                .build()
+                .adapter(PathPropRequest::class.java)
+    }
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt]: You need the distance adapters (see [here](https://github.com/OpenRailAssociation/osrd/blob/dev/core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlockResponse.kt#L98-L105))

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropResult.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class PathPropResult(
+    val slopes: RangeValues<Double>,
+    val gradients: RangeValues<Double>,
+    val electrifications: RangeValues<Electrification>,
+    var geometry: RJSLineString
+) {
+
+    companion object {
+        val adapterResult: JsonAdapter<PathPropResult> =
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropResult.kt]: Same, missing distance adapter

The polymorphism on electrification requires a little extra work as well (to define which field defines the type). See [here](https://github.com/OpenRailAssociation/osrd/blob/dev/core/src/main/kotlin/fr/sncf/osrd/api/api_v2/pathfinding/PathfindingBlockResponse.kt#L87-L96) for an example

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropResultConverter.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    // If there are no supported electrifications, return NonElectrified along the whole path
+    if (rsSupportedElectrifications.isEmpty())
+        return RangeValues(listOf(Distance.ZERO, pathProperties.getLength().distance), listOf(Electrification.NonElectrified))
+    val electrifications = makeElectrificationMap(pathProperties.getElectrification(), rsSupportedElectrifications)
+    val neutralSections = makeElectrificationMap(pathProperties.getNeutralSections())
+    val combinedEntries = (electrifications + neutralSections).sortedBy { it.lower }
+    val mergedEntries = mutableListOf<DistanceRangeMap.RangeMapEntry<Electrification>>()
+    var i = 0
+    var currentLower = Distance.ZERO
+    while (i < combinedEntries.size) {
```
Comment (eckter) [inline core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropResultConverter.kt]: With that much complexity, we could use `com.google.common.collect.RangeMap`. It has a lot more utility functions builtin, in particular methods to merge different maps with easy ways to handle overlaps. We could convert to `RangeMap`, do all that processing, and then convert back to a `RangeValues`. It would be a little less efficient, but I'd be more confident in the reliability of that code

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package fr.sncf.osrd.sim_infra.api
 
 import fr.sncf.osrd.geom.LineString
+import fr.sncf.osrd.sim_infra.impl.new_impl.OperationalPointPartProps
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt]: OperationalPointPartProps should be part of the `api` folder, as it is returned by getOperationalPointPartProps. don't let the impl leak into the API.

File: core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt [hunk 6]
```diff
@@ -756,11 +756,7 @@ fun parseRJSInfra(rjsInfra: RJSInfra): RawInfra {
             if (distinctParts.contains(opPart)) continue
 
             distinctParts.add(opPart)
-            builder.operationalPointPart(
-                operationalPoint.id,
-                opPart.track,
-                Offset(opPart.position.meters)
-            )
+            builder.operationalPointPart(operationalPoint, opPart)
```
Comment (multun) [inline core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt]: the builder API is not tied to RailJSON. please parse the input in the RailJSON parser, not the RawInfra builder

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 7]
```diff
@@ -522,18 +524,33 @@ class RawInfraFromRjsBuilder {
     }
 
     fun operationalPointPart(
-        operationalPointName: String,
-        trackSectionName: String,
-        trackSectionOffset: Offset<TrackSection>
+        op: RJSOperationalPoint,
+        opPart: RJSOperationalPointPart
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt]: please do not parse RailJSON in the builder

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt [hunk 8]
```diff
@@ -59,11 +60,17 @@ interface TrackProperties {
...
     ): Offset<TrackChunk>
 
-    fun getOperationalPointPartName(operationalPoint: OperationalPointPartId): String
+    fun getOperationalPointPartOpId(operationalPointPart: OperationalPointPartId): String
+
+    fun getOperationalPointPartProps(
+        operationalPointPart: OperationalPointPartId
+    ): OperationalPointPartProps
+
+    fun getOperationalPointPartKp(operationalPointPart: OperationalPointPartId): String?
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt]: I'm aware that the current data model of RawInfra does not quite match the data model of RailJSON: in the new infra API, there are just OperationalPointPart, no OperationalPoint.

That's because merging both types is, as far as I know, the eventual target for RailJSON: parent operational points are only useful as repositories for shared metadata. the issue is: shared using what criteria? we would rather have a single type, and have one operational point that's actually a point than go on with the current mess.

But right now, there's a mismatch between the two APIs. Here are some ways it could be solved:
- the eventual target (removing the OperationalPointPart object and having one OperationalPoint per track location) is too costly to do now
- adopt the RailJSON model and migrate later
- perform the de-normalization step (adding the metadata of operational points into parts) on import

I believe the last option has the best long-term viability, and is also the easiest to implement. However, I don't really like how it's currently done:
- despite having performed de-normalization, there is a separate call for the kilometric point (getOperationalPointPartKp)
- RailJSON extensions leak into the SimInfra API, even though they could really be opaque (data is just passed around, not read)

I would instead suggest the following:

```suggestion
    // returns a unique string identifier for this object
    fun getOperationalPointPartName(operationalPointPart: OperationalPointPartId): String

    fun getOperationalPointPartProps(
        operationalPointPart: OperationalPointPartId
    ): Map<String, String>
```

during RailJSON parsing, the name of the operationalpointpart would be built as follows: `${operationalPoint.name}.${operationalPointPart.index}`

Props would be a generic metadata storage facility, which could be later used for querying objects having some property value (indices can be built lazily). When returning operational points on the path, properties can be parsed back for editoast consumption.

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
+import fr.sncf.osrd.api.api_v2.pathfinding.TrackRange
+import fr.sncf.osrd.utils.json.DistanceAdapterFactory
+import fr.sncf.osrd.utils.json.OffsetAdapterFactory
+
+class PathPropRequest(
+    @Json(name = "track_section_ranges") val trackSectionRanges: List<TrackRange>,
+    @Json(name = "rolling_stock_supported_electrifications")
+    val rollingStockSupportedElectrifications: List<String>,
```
Comment (multun) [inline core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt]:
```suggestion
```

I would rather get the client to perform this filtering pass. it's not a lot of data, and it enables figuring out why a rolling stock can't go somewhere, in addition to whether it can

it would also make the API clearer: after all, it's an API about path properties, not rolling stocks yet

**Review by multun — CHANGES_REQUESTED** [review]

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt (hunk 8 above)
Comment (Erashin) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/TrackProperties.kt]: I separated kp from the rest because it's an operationalPoint prop, whereas the others were operationalPointPartProps. This would be a bit more complex, but would globally work out, so why not.

**Review by younesschrifi — APPROVED** [review]
LGTM ! Reviewed editoast part :)

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt (hunk 9 above)
Comment (flomonster) [reply core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt]: We can already figure out why a rolling stock can't go somewhere, using the pathfinding endpoint.

File: core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt [hunk 10]
```diff
@@ -754,12 +766,33 @@ fun parseRJSInfra(rjsInfra: RJSInfra): RawInfra {
...
             if (distinctParts.contains(opPart)) continue
-
             distinctParts.add(opPart)
+
+            val operationalPointId = operationalPoint.id
+            val trackSectionName = opPart.track
+            val trackSectionOffset = Offset<TrackSection>(opPart.position.meters)
+            val props = mutableMapOf<String, String>()
+            props["trackSectionName"] = trackSectionName
+            props["position"] = opPart.position.toString()
```
Comment (multun) [inline core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt]: why is that needed? isn't it already available using other APIs?

File: core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt (hunk 9 above)
Comment (multun) [reply core/src/main/kotlin/fr/sncf/osrd/api/api_v2/path_properties/PathPropRequest.kt]: We discussed the matter and came to the conclusion that it's best to remove this argument, as:
- it allows for easier caching, using only (infra, path)
- it makes the API less error prone, as the client can't mistake unsupported electrified zones for non-electrified zones

We also considered changing the result API to make the second point clearer, but decided against it because of the first point

File: core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt (hunk 10 above)
Comment (Erashin) [reply core/kt-osrd-rjs-parser/src/main/kotlin/fr/sncf/osrd/RawInfraRJSParser.kt:776]: `trackSectionName` is indeed unneeded. However, we want to keep `position` as a double, so i need to keep it there.

**Review by multun — APPROVED** [review]

File: editoast/src/core/v2/path_properties.rs [hunk 11]
```diff
@@ -0,0 +1,84 @@
+use editoast_common::geometry::GeoJsonLineString;
+use editoast_common::Identifier;
+use editoast_schemas::infra::OperationalPointExtensions;
+use editoast_schemas::infra::OperationalPointPart;
+use serde::Deserialize;
+use serde::Serialize;
+use utoipa::ToSchema;
+
+use crate::core::{AsCoreRequest, Json};
+use crate::views::v2::path::TrackRange;
```
Comment (leovalais) [inline editoast/src/core/v2/path_properties.rs]: Not blocking or anything but I'd like to avoid having `editoast::core` depend on `views` (that's extra work for the split in crates ; one struct is fine though :))

**Review by leovalais — APPROVED** [review]
LGTM

File: editoast/src/core/v2/path_properties.rs (hunk 11 above)
Comment (flomonster) [reply editoast/src/core/v2/path_properties.rs]: Good catch
